pkg/controller/common.go

```diff
@@ -1653,33 +1644,43 @@ func (c *SharedState) cleanUpAzVolumeAttachmentByNode(ctx context.Context, azDri
 			},
 			false)
 	}
-	return attachments, nil
+	return attachments.Items, nil
 }
 
 func (c *SharedState) cleanUpAzVolumeAttachments(ctx context.Context, attachments []diskv1beta1.AzVolumeAttachment, cleanUp cleanUpMode, caller operationRequester) error {
 	for _, attachment := range attachments {
+		var patchRequired bool
 		patched := attachment.DeepCopy()
+
 		if patched.Annotations == nil {
 			patched.Annotations = map[string]string{}
 		}
-
 		// if caller is azdrivernode, don't append cleanup annotation
-		if caller != azdrivernode {
+		if caller != azdrivernode && !metav1.HasAnnotation(patched.ObjectMeta, consts.CleanUpAnnotation) {
+			patchRequired = true
 			patched.Annotations[consts.CleanUpAnnotation] = string(caller)
 		}
 		// replica attachments should always be detached regardless of the cleanup mode
 		if cleanUp == detachAndDeleteCRI || patched.Spec.RequestedRole == diskv1beta1.ReplicaRole {
-			patched.Annotations[consts.VolumeDetachRequestAnnotation] = string(caller)
+			if !metav1.HasAnnotation(patched.ObjectMeta, consts.VolumeDetachRequestAnnotation) {
+				patchRequired = true
+				patched.Annotations[consts.VolumeDetachRequestAnnotation] = string(caller)
+			}
 		}
```

Comment:
It seems like the previous two if statements could be enclosed by a single check for the absence of the annotation rather than checking twice (once for each case).